File: frontend_manager.py
import traceback

from fastapi import WebSocket
from typing import Set, Optional, Dict, Tuple
import asyncio
import uuid


class FrontendConnectionManager:
    def __init__(self, onebot_manager=None, req_backend_handlers: dict = None):
        self.active_connections: Set[WebSocket] = set()
        self.onebot_manager = onebot_manager
        # 存储正在执行的 action 任务 (echo -> (websocket, Task))
        self.pending_action_tasks: Dict[str, Tuple[WebSocket, asyncio.Task]] = {}
        # req_backend 处理器映射: endpoint -> handler_function(params) -> dict
        self.req_backend_handlers = req_backend_handlers or {}
        # 存储每个连接的 self_id（前端选择的bot）
        self.connection_self_id: Dict[WebSocket, Optional[str]] = {}

    async def connect(self, websocket: WebSocket, self_id: Optional[str] = None):
        """处理新的WebSocket连接"""
        await websocket.accept()
        self.active_connections.add(websocket)
        if self_id:
            self.connection_self_id[websocket] = self_id
        print(f"新前端连接，当前连接数: {len(self.active_connections)}, self_id: {self_id}")

        try:
            while True:
                data = await websocket.receive_json()
                await self.process_message(websocket, data)
        except Exception as e:
            print(f"连接异常: {e}")
        finally:
            self.disconnect(websocket)

    def disconnect(self, websocket: WebSocket):
        """处理连接断开"""
        # 取消该连接相关的所有 pending action 任务
        for echo, (ws, task) in list(self.pending_action_tasks.items()):
            if ws is websocket and not task.done():
                task.cancel()
                self.pending_action_tasks.pop(echo, None)

        if websocket in self.active_connections:
            self.active_connections.remove(websocket)
        self.connection_self_id.pop(websocket, None)
        print(f"前端断开连接，剩余连接数: {len(self.active_connections)}")

    async def process_message(self, websocket: WebSocket, message: dict):
        """处理来自前端的消息"""
        msg_type = message.get("type", "")

        if msg_type == "send_action":
            echo = message.get("echo", str(uuid.uuid4()))
            # 将 action 作为后台任务运行，以便可以取消
            task = asyncio.create_task(
                self.handle_send_action(websocket, message, echo)
            )
            self.pending_action_tasks[echo] = (websocket, task)
            task.add_done_callback(lambda _: self.pending_action_tasks.pop(echo, None))
        elif msg_type == "cancel_action":
            await self.handle_cancel_action(websocket, message)
        elif msg_type == "req_backend":
            echo = message.get("echo", str(uuid.uuid4()))
            task = asyncio.create_task(
                self.handle_req_backend(websocket, message, echo)
            )
            self.pending_action_tasks[echo] = (websocket, task)
            task.add_done_callback(lambda _: self.pending_action_tasks.pop(echo, None))
        else:
            print("收到消息:", message)
            print(websocket)

    async def handle_cancel_action(self, websocket: WebSocket, message: dict):
        """处理前端发来的 cancel_action 请求，取消正在执行的 action"""
        echo = message.get("echo", "")
        if not echo:
            return

        entry = self.pending_action_tasks.pop(echo, None)
        if entry:
            ws, task = entry
            if not task.done():
                task.cancel()
                print(f"取消 action (echo: {echo})")

    async def handle_send_action(self, websocket: WebSocket, message: dict, echo: str):
        """处理前端发来的 send_action 请求，转发给 OneBot 并返回结果"""
        try:
            if not self.onebot_manager:
                await websocket.send_json({
                    "type": "send_action_response",
                    "echo": echo,
                    "status": "error",
                    "message": "OneBot manager not available"
                })
                return

            action = message.get("action", "")
            params = message.get("params", {})
            timeout = params.get("timeout", 60)

            # 获取该连接对应的 self_id，优先使用连接时指定的 self_id
            self_id = self.connection_self_id.get(websocket)

            # print(f"前端请求 action: {action}, params: {params}, self_id: {self_id}")

            try:
                result = await self.onebot_manager.call_action(action, params, self_id, timeout)
                # print(f"action 响应: {action} -> {result}")

                response = {
                    "type": "send_action_response",
                    "echo": echo,
                    "status": "ok",
                    "data": result
                }
            except asyncio.CancelledError:
                # action 被取消，发送取消响应
                print(f"action 已取消: {action} (echo: {echo})")
                response = {
                    "type": "send_action_response",
                    "echo": echo,
                    "status": "cancelled",
                    "message": "Action cancelled by user"
                }
            except Exception as e:
                print(f"action 失败: {action} -> {e}")
                response = {
                    "type": "send_action_response",
                    "echo": echo,
                    "status": "error",
                    "message": str(e)
                }

            await websocket.send_json(response)
        except asyncio.CancelledError:
            # 任务被取消时的清理
            raise

    async def handle_req_backend(self, websocket: WebSocket, message: dict, echo: str):
        """处理前端发来的 req_backend 请求，调用本地注册的处理器并返回结果"""
        try:
            endpoint = message.get("endpoint", "")
            params = message.get("params", {})

            if endpoint not in self.req_backend_handlers:
                await websocket.send_json({
                    "type": "req_backend_response",
                    "echo": echo,
                    "status": "error",
                    "message": f"Unknown endpoint: {endpoint}"
                })
                return

            handler = self.req_backend_handlers[endpoint]
            try:
                result = await handler(params)
                response = {
                    "type": "req_backend_response",
                    "echo": echo,
                    "status": "ok",
                    "data": result
                }
            except asyncio.CancelledError:
                print(f"req_backend 已取消: {endpoint} (echo: {echo})")
                response = {
                    "type": "req_backend_response",
                    "echo": echo,
                    "status": "cancelled",
                    "message": "Request cancelled by user"
                }
            except Exception as e:
                print(f"req_backend 失败: {endpoint} -> {e}")
                traceback.print_exc()
                response = {
                    "type": "req_backend_response",
                    "echo": echo,
                    "status": "error",
                    "message": str(e)
                }

            await websocket.send_json(response)
        except asyncio.CancelledError:
            raise

    async def broadcast(self, message: dict):
        """广播消息给所有连接的前端"""
        if len(self.active_connections) > 0:
            await asyncio.gather(
                *[connection.send_json(message) for connection in self.active_connections]
            )

File: server/src/frontend_manager.ts
import type { WebSocket } from '@fastify/websocket';
import type { FastifyRequest } from 'fastify';
import { randomUUID } from 'node:crypto';
import { OneBotConnectionManager } from './onebot_manager.js';

/**
 * 前端 WebSocket 连接管理器
 */
export class FrontendConnectionManager {
  activeConnections: Set<WebSocket> = new Set();
  onebotManager: OneBotConnectionManager | null;
  // 存储正在执行的 action 任务 (echo -> { websocket, abort })
  pendingActionTasks: Map<string, { websocket: WebSocket; abort: AbortController }> = new Map();
  // req_backend 处理器映射: endpoint -> handler_function(params) -> dict
  reqBackendHandlers: Map<string, (params: Record<string, unknown>) => Promise<unknown>> = new Map();
  // 存储每个连接的 self_id（前端选择的bot，int 类型）
  connectionSelfId: Map<WebSocket, number | null> = new Map();

  constructor(onebotManager: OneBotConnectionManager | null = null, reqBackendHandlers: Record<string, (params: Record<string, unknown>) => Promise<unknown>> = {}) {
    this.onebotManager = onebotManager;
    for (const [ key, handler ] of Object.entries(reqBackendHandlers)) {
      this.reqBackendHandlers.set(key, handler);
    }
  }

  async connect(websocket: WebSocket, request: FastifyRequest, selfId: string | null = null): Promise<void> {
    /* 处理新的WebSocket连接 */
    // @fastify/websocket 已自动 accept 连接
    this.activeConnections.add(websocket);
    if (selfId) {
      this.connectionSelfId.set(websocket, Number(selfId));
    }
    console.log(`新前端连接，当前连接数: ${this.activeConnections.size}, self_id: ${selfId}`);

    // 使用事件监听方式处理消息（ws 库事件驱动模式）
    const onMessage = (rawData: Buffer | string) => {
      try {
        const messageStr = typeof rawData === 'string' ? rawData : rawData.toString();
        const data = JSON.parse(messageStr) as Record<string, unknown>;
        // 异步处理但不阻塞事件循环
        this.processMessage(websocket, data).catch((err) => {
          console.error(`WebSocket process message error: ${err}`);
        });
      } catch (err) {
        console.log(`WebSocket message parse error: ${err}`);
      }
    };

    const onClose = () => {
      this.disconnect(websocket);
    };

    const onError = (err: Error) => {
      console.log(`WebSocket error: ${err}`);
    };

    websocket.on('message', onMessage);
    websocket.on('close', onClose);
    websocket.on('error', onError);
  }

  disconnect(websocket: WebSocket): void {
    /* 处理连接断开 */
    // 取消该连接相关的所有 pending action 任务
    for (const [ echo, entry ] of this.pendingActionTasks) {
      if (entry.websocket === websocket) {
        entry.abort.abort();
        this.pendingActionTasks.delete(echo);
      }
    }

    this.activeConnections.delete(websocket);
    this.connectionSelfId.delete(websocket);
    console.log(`前端断开连接，剩余连接数: ${this.activeConnections.size}`);
  }

  async processMessage(websocket: WebSocket, message: Record<string, unknown>): Promise<void> {
    /* 处理来自前端的消息 */
    const msgType = String(message['type'] ?? '');

    if (msgType === 'send_action') {
      const echo = String(message['echo'] ?? randomUUID());
      const abortController = new AbortController();
      this.pendingActionTasks.set(echo, { websocket, abort: abortController });

      try {
        await this.handleSendAction(websocket, message, echo, abortController.signal);
      } finally {
        this.pendingActionTasks.delete(echo);
      }
    } else if (msgType === 'cancel_action') {
      await this.handleCancelAction(websocket, message);
    } else if (msgType === 'req_backend') {
      const echo = String(message['echo'] ?? randomUUID());
      const abortController = new AbortController();
      this.pendingActionTasks.set(echo, { websocket, abort: abortController });

      try {
        await this.handleReqBackend(websocket, message, echo, abortController.signal);
      } finally {
        this.pendingActionTasks.delete(echo);
      }
    } else {
      console.log('收到消息:', message);
      console.log(websocket);
    }
  }

  async handleCancelAction(websocket: WebSocket, message: Record<string, unknown>): Promise<void> {
    /* 处理前端发来的 cancel_action 请求，取消正在执行的 action */
    const echo = String(message['echo'] ?? '');
    if (!echo) return;

    const entry = this.pendingActionTasks.get(echo);
    if (entry) {
      entry.abort.abort();
      this.pendingActionTasks.delete(echo);
      console.log(`取消 action (echo: ${echo})`);
    }
  }

  async handleSendAction(websocket: WebSocket, message: Record<string, unknown>, echo: string, signal: AbortSignal): Promise<void> {
    /* 处理前端发来的 send_action 请求，转发给 OneBot 并返回结果 */
    try {
      if (!this.onebotManager) {
        await this.sendJson(websocket, {
          type: 'send_action_response',
          echo,
          status: 'error',
          message: 'OneBot manager not available',
        });
        return;
      }

      const action = String(message['action'] ?? '');
      const params = (message['params'] as Record<string, unknown>) ?? {};
      const timeout = Number(params['timeout'] ?? 60);

      // 获取该连接对应的 self_id，优先使用连接时指定的 self_id
      const selfId = this.connectionSelfId.get(websocket) ?? null;

      try {
        const result = await this.onebotManager.callAction(action, params, selfId, timeout);

        const response = {
          type: 'send_action_response',
          echo,
          status: 'ok',
          data: result,
        };
        await this.sendJson(websocket, response);
      } catch (err) {
        if (signal.aborted) {
          console.log(`action 已取消: ${action} (echo: ${echo})`);
          await this.sendJson(websocket, {
            type: 'send_action_response',
            echo,
            status: 'cancelled',
            message: 'Action cancelled by user',
          });
        } else {
          console.log(`action 失败: ${action} -> ${err}`);
          await this.sendJson(websocket, {
            type: 'send_action_response',
            echo,
            status: 'error',
            message: String(err),
          });
        }
      }
    } catch (err) {
      // 任务被取消时的清理
      throw err;
    }
  }

  async handleReqBackend(websocket: WebSocket, message: Record<string, unknown>, echo: string, signal: AbortSignal): Promise<void> {
    /* 处理前端发来的 req_backend 请求，调用本地注册的处理器并返回结果 */
    try {
      const endpoint = String(message['endpoint'] ?? '');
      const params = (message['params'] as Record<string, unknown>) ?? {};

      // 注入该连接对应的 self_id 到 params 中
      const selfId = this.connectionSelfId.get(websocket);
      if (selfId !== undefined && selfId !== null) {
        params['self_id'] = selfId;
      }

      const handler = this.reqBackendHandlers.get(endpoint);
      if (!handler) {
        await this.sendJson(websocket, {
          type: 'req_backend_response',
          echo,
          status: 'error',
          message: `Unknown endpoint: ${endpoint}`,
        });
        return;
      }

      try {
        const result = await handler(params);
        await this.sendJson(websocket, {
          type: 'req_backend_response',
          echo,
          status: 'ok',
          data: result,
        });
      } catch (err) {
        if (signal.aborted) {
          console.log(`req_backend 已取消: ${endpoint} (echo: ${echo})`);
          await this.sendJson(websocket, {
            type: 'req_backend_response',
            echo,
            status: 'cancelled',
            message: 'Request cancelled by user',
          });
        } else {
          console.log(`req_backend 失败: ${endpoint} -> ${err}`);
          console.error(err);
          await this.sendJson(websocket, {
            type: 'req_backend_response',
            echo,
            status: 'error',
            message: String(err),
          });
        }
      }
    } catch (err) {
      throw err;
    }
  }

  async broadcast(message: Record<string, unknown>): Promise<void> {
    /* 广播消息给所有连接的前端 */
    if (this.activeConnections.size > 0) {
      await Promise.all(
        [ ...this.activeConnections ].map((conn) => this.sendJson(conn, message).catch(() => {
        })),
      );
    }
  }

  async broadcastToSelfId(selfId: number, message: Record<string, unknown>): Promise<void> {
    /* 仅广播消息给指定 self_id 对应的前端连接 */
    const targetConnections = [ ...this.connectionSelfId.entries() ]
      .filter(([ ws, sid ]) => sid === selfId && this.activeConnections.has(ws))
      .map(([ ws ]) => ws);

    if (targetConnections.length > 0) {
      await Promise.all(
        targetConnections.map((conn) => this.sendJson(conn, message).catch(() => {
        })),
      );
    }
  }

  private async sendJson(websocket: WebSocket, data: Record<string, unknown>): Promise<void> {
    websocket.send(JSON.stringify(data));
  }
}